sdk/core/azure-core-amqp/src/amqp/connection.cpp

```diff
@@ -556,13 +589,74 @@ namespace Azure { namespace Core { namespace Amqp { namespace _detail {
     return false;
   }
 
-  std::string ConnectionImpl::GetSecurityToken(
+  // Ensure that we have a token for the provided audience.
+  // If we don't, authenticate the audience with the service using the provided session.
+  // Note that the granularity of
+  Credentials::AccessToken ConnectionImpl::AuthenticateAudience(
+      std::shared_ptr<SessionImpl> session,
       std::string const& audience,
-      Azure::Core::Context const& context) const
+      Azure::Core::Context const& context)
   {
     if (GetCredential())
     {
-      if (m_tokenStore.find(audience) == m_tokenStore.end())
+      std::string audienceUrl = audience;
+      if (m_options.EnableTrace)
+      {
+        Log::Stream(Logger::Level::Verbose) << "Authenticate connection for audience " << audience;
+      }
+      // If the audience looks like a URL for AMQP, AMQPS, or SB, we can use the URL as provided.
+      if ((audience.find("amqps://") != 0) && (audience.find("amqp://") != 0)
+          && (audience.find("sb://") != 0))
+      {
+        audienceUrl = "amqps://" + GetHost();
+        // The provided audience may begin with a /, if not, we need to add the separator.
+        if (audience.front() != '/')
+        {
+          audienceUrl += "/";
+        }
+        audienceUrl += audience;
+        if (m_options.EnableTrace)
+        {
+          Log::Stream(Logger::Level::Verbose)
+              << "Initial audience is not URL, using " << audienceUrl;
+        }
+      }
+
+      std::unique_lock<std::mutex> lock(m_tokenMutex);
+      // If we have authenticated this audience, we're done and can return success.
+      auto token = m_tokenStore.find(audienceUrl);
+      if (token != m_tokenStore.end())
+      {
+        if (m_options.EnableTrace)
+        {
+          Log::Stream(Logger::Level::Verbose) << "Using cached token for " << audienceUrl;
+        }
+        return token->second;
+      }
+      // We've not authenticated this audience.
+      // Authenticate it with the server
+
+      if (m_options.EnableTrace)
+      {
+        Log::Stream(Logger::Level::Verbose)
+            << "No cached token for " << audienceUrl << ", Authenticating.";
+      }
+      // Azure::Core::Amqp::_internal::SessionOptions sessionOptions;
+      // sessionOptions.InitialIncomingWindowSize = std::numeric_limits<int32_t>::max();
+      // sessionOptions.InitialOutgoingWindowSize = std::numeric_limits<uint16_t>::max();
+
+      // auto authenticationSession{std::make_shared<SessionImpl>(
+      //     shared_from_this(), sessionOptions, nullptr)};
```

Comment:
commented code

Reply:
This is code which is blocked because of uAMQP bugs that need to have a vendored version of uAMQP.